### PR TITLE
Automated cherry pick of #95664: Don't record events in goroutines

### PR DESCRIPTION
**Pull request by dims**
Cherry pick of #95664 on release-1.20.

#95664: Don't record events in goroutines

For details on the cherry pick process, see the [cherry pick requests](https://git.k8s.io/community/contributors/devel/sig-release/cherry-picks.md) page.

### CODE REVIEW

**Review by hasheddan — COMMENTED** [review]
/retest

**Comment by vincepri** [comment]
LGTM

Thank you @dims !

**Comment by hasheddan** [comment]
@dims is this applicable to 1.19 / 1.18?

**Comment by dims** [comment]
@hasheddan the request was to cherry-pick to 1.20 ( https://github.com/kubernetes/kubernetes/pull/95664#issuecomment-767003021 ) 

@liggitt @DirectXMan12 do we need this in 1.19 / 1.18?

**Comment by liggitt** [comment]
it's not clear to me whether this should be cherry-picked or not.

how long has this issue existed, what is the impact of the issue (non-clean shutdown?), and what is the risk of the fix (increased memory usage, etc)?

**Comment by vincepri** [comment]
@liggitt For 1.20 (I haven't been able to reproduce on earlier versions) and Controller Runtime v0.8.x, shutdown panics more often than not. The consequence is that sometimes leader election might not cleanup properly on exit and tests flake more often than not.

We'll adopt 1.21 soon in CR, although I wanted to make sure folks aren't getting caught off guard when using 0.8.x branch which is based on 1.20.

File: staging/src/k8s.io/client-go/tools/record/event.go [hunk 1]
```diff
@@ -338,11 +338,14 @@ func (recorder *recorderImpl) generateEvent(object runtime.Object, annotations m
 	event := recorder.makeEvent(ref, annotations, eventtype, reason, message)
 	event.Source = recorder.source
 
-	go func() {
-		// NOTE: events should be a non-blocking operation
```
Comment (liggitt) [inline staging/src/k8s.io/client-go/tools/record/event.go:342]: for backporting, would it be less invasive to switch this to something like this:

```go
defer func(){
  if err := recover(); err != nil {
    utilruntime.HandleError(err)
  }
}()
```

I'd like more soak time on the doubling of the event queue before backporting

Comment (vincepri) [reply staging/src/k8s.io/client-go/tools/record/event.go:342]: cc @DirectXMan12

**Pull request closed without merging**